### DISCUSSION TITLE
Apariencia/Diseño

### DISCUSSION THREAD

**Discussion by oleksis** [oleksis/picta-dl-gui#3] [Ideas]
Aquí podemos debatir sobre la **Apariencia y Animaciones** (**Themes**) para la GUI en diferentes plataformas:

<img src="https://user-images.githubusercontent.com/25087943/102360783-36f89280-3f91-11eb-8ddf-7a622d361502.png"   width=200 />

<img src="https://user-images.githubusercontent.com/25087943/102360805-3d870a00-3f91-11eb-9725-cdc9748ab1cd.png"   width=200 />

También subir Mocking o ScreenShots de las ideas 💡

**Comment by oleksis** [comment]
🔳Puedes probar Picta-DL-GUI aplicando temas oscuros en Windows por ejemplo :https://blogs.windows.com/windowsexperience/2019/04/01/windows-10-tip-dark-theme-in-file-explorer/🔲

Asi es como se ven los temas en Windows:

![Picta-dl_GUI-Dark](https://user-images.githubusercontent.com/44526468/102517350-ebf88100-405d-11eb-91b2-dedb07e8e742.png)

![Picta-dl_GUI-Light](https://user-images.githubusercontent.com/44526468/102517382-f6b31600-405d-11eb-994d-c4911cbe9d1b.png)

**Comment by rsanjuan87** [comment]
Prueben esto y me dicen q les parece
```
#ifdef Q_OS_WIN
    QSettings settings("HKEY_CURRENT_USER\\Software\\Microsoft\\Windows\\CurrentVersion\\Themes\\Personalize",QSettings::NativeFormat);
    if(settings.value("AppsUseLightTheme")==0){
        qApp->setStyle(QStyleFactory::create("Fusion"));
        QPalette darkPalette;
        QColor darkColor = QColor(10,10,10);
        QColor disabledColor = QColor(127,127,127);
        darkPalette.setColor(QPalette::Window, darkColor);
        darkPalette.setColor(QPalette::WindowText, Qt::white);
        darkPalette.setColor(QPalette::Base, QColor(18,18,18));
        darkPalette.setColor(QPalette::AlternateBase, darkColor);
        darkPalette.setColor(QPalette::ToolTipBase, Qt::white);
        darkPalette.setColor(QPalette::ToolTipText, Qt::white);
        darkPalette.setColor(QPalette::Text, Qt::white);
        darkPalette.setColor(QPalette::Disabled, QPalette::Text, disabledColor);
        darkPalette.setColor(QPalette::Button, darkColor);
        darkPalette.setColor(QPalette::ButtonText, Qt::white);
        darkPalette.setColor(QPalette::Disabled, QPalette::ButtonText, disabledColor);
        darkPalette.setColor(QPalette::BrightText, Qt::red);
        darkPalette.setColor(QPalette::Link, QColor(42, 130, 218));

        darkPalette.setColor(QPalette::Highlight, QColor(42, 130, 218));
        darkPalette.setColor(QPalette::HighlightedText, Qt::black);
        darkPalette.setColor(QPalette::Disabled, QPalette::HighlightedText, disabledColor);

        qApp->setPalette(darkPalette);
        QFile file(":/winApp.qss");
        file.open(QFile::ReadOnly);
        QString styleSheet = QString::fromLatin1(file.readAll());
        styleSheet=styleSheet.replace("#001133", darkColor.name());
        qApp->setStyleSheet(styleSheet);
    }
#endif
```

[winApp.qss](https://github.com/oleksis/picta-dl-gui/files/5711071/winApp.qss.txt)

**Reply by oleksis** [reply]
Si puedes adicionar los cambios usando QSS pues adicioné el tema oscuro directamente sin usar recursos

**Reply by rsanjuan87** [reply]
use el resurso para no tener un texto inmenso en el code
asi se mantiene mas limpio y facil de editar
PD: modifiquen a consideracion

**Reply by rsanjuan87** [reply]
<img width="642" alt="preview" src="https://user-images.githubusercontent.com/25087943/102530292-da27d580-407f-11eb-8700-83ae69371a2e.png">

**Comment by rsanjuan87** [comment]
Ajustes al `QCommandLinkButton`
[winApp.qss](https://github.com/oleksis/picta-dl-gui/files/5711470/winApp.qss.txt)

**Reply by oleksis** [reply]
OK Descargando el qss y adicionando los cambios!

**Reply by rsanjuan87** [reply]
siempre verificar y probar

**Reply by rsanjuan87** [reply]
@oleksis si puedes revisa q veo q el QCommandLinkButton se fue en el qss con un min-with muy grande

**Reply by rsanjuan87** [reply]
remover 
```
 min-width: 30ex;
 min-width: 30ex;
```

del [winApp.qss](https://github.com/oleksis/picta-dl-gui/files/5711470/winApp.qss.txt)

**Reply by oleksis** [reply]
Corregido!

**Reply by Spheres-cu** [reply]
Saludos, estoy preparando nuevos iconos en escala de grises para los temas claros-oscuros

**Reply by oleksis** [reply]
> Saludos, estoy preparando nuevos iconos en escala de grises para los temas claros-oscuros

Que bien 👍🏾 Pienso que es casi lo último que faltaría para lanzar una versión "Christmas" claro-oscuro como regalo de fin de Año 🎁

**Reply by Spheres-cu** [reply]
Ya hice los cambios, ahora falta las animaciones, hay cambios en el mainwindows.ui para ajustar el icono de delete

**Comment by Spheres-cu** [comment]
Aquí una muestra de los cambios
![Picta-dl_GUI_modo-oscuro-claro](https://user-images.githubusercontent.com/72468068/102566352-600e4580-40ad-11eb-8d95-891ab2b6c82f.png)

**Reply by oleksis** [reply]
Me gustan los cambios! Se ven más finos y limpios 😍

**Comment by Spheres-cu** [comment]
![picta-dl-config-dark](https://user-images.githubusercontent.com/72468068/102566981-8f718200-40ae-11eb-9058-459c6828dc8b.png)
También cambios en la imagen de la ayuda

**Reply by oleksis** [reply]
Ok 👍🏾Cuando estén listos los iconos/animaciones envíalos para integrarlo/probar y valorar cuando hacemos el release con estos buenos cambios en el Look&Feel

**Reply by Spheres-cu** [reply]
Estoy teniendo problemas con las animaciones, en el modo oscuro salen distorsionadas

**Reply by oleksis** [reply]
> Estoy teniendo problemas con las animaciones, en el modo oscuro salen distorsionadas

Será por la calidad del gif o reajustar la GUI esos componentes (QCommandLinkButtons) . Algunos estilos se aplican en winApp.qss al establecer el tema oscuro.

**Comment by Spheres-cu** [comment]
Me esta saliendo desde el inicio un warning por los estilos nuevos: QIODevice::read (QFile, ":\Styles\winApp.qss"): device not open

**Reply by oleksis** [reply]
Te fijaste en hacer un Fetch/Pull al repositorio para actualizar los cambios. Chequea que existe la Carpeta en Resources

**Reply by oleksis** [reply]
> Me esta saliendo desde el inicio un warning por los estilos nuevos: QIODevice::read (QFile, ":\Styles\winApp.qss"): device not open

Si tienes la carpeta Resources/Styles/winApp.qss chequea/adiciona/modifica el proyecto para que utilice winApp.qss como recurso

**Reply by Spheres-cu** [reply]
La verdad que no queda muy bien con ese estilo
![1](https://user-images.githubusercontent.com/72468068/102577226-a66f9e80-40c5-11eb-85e9-e5d86491d4f7.png)

**Reply by oleksis** [reply]
Porque los botones están muy grandes?  Parece que no se están aplicando todos los estilos. La imagen que te comenté aquí: https://github.com/oleksis/picta-dl-gui/discussions/3#discussioncomment-220941 fíjate que los botones lucen diferentes.

**Reply by oleksis** [reply]
> La verdad que no queda muy bien con ese estilo

Etiquete la version que te mustro en las capturas 🏷 https://github.com/oleksis/picta-dl-gui/archive/0.12.4.zip

**Comment by Spheres-cu** [comment]
Tambien con el tema oscuro no funciona el borrar ni el procesar URL cuando terminas la descarga

**Reply by oleksis** [reply]
He probado descarga Videos y Limpiar/Borrar con los últimos cambios y me funciona

**Reply by oleksis** [reply]
> Tambien con el tema oscuro no funciona el borrar ni el procesar URL cuando terminas la descarga

Te envío una captura. De paso a veces el porciento de descarga cuando finaliza no reporta el 100% 

![D966CDBD-1F0E-4D18-A7D6-33A531024639](https://user-images.githubusercontent.com/44526468/102576315-35c78280-40c3-11eb-9333-a187e0a7cbe4.jpeg)

**Reply by Spheres-cu** [reply]
> He probado descarga Videos y Limpiar/Borrar con los últimos cambios y me funciona

Si ya el problema se le quitó, parece que era una función nueva que tuve que hacer para la animación con PNG y no con Gif ya que los gif en modo oscuro se ven mal, estoy tratando de hacer una función de animación para corregir eso, de lo contrario un gif para modo oscuro y otro para modo claro

**Reply by Spheres-cu** [reply]
> > Tambien con el tema oscuro no funciona el borrar ni el procesar URL cuando terminas la descarga
> 
> Te envío una captura. De paso a veces el porciento de descarga cuando finaliza no reporta el 100%
> 
> ![D966CDBD-1F0E-4D18-A7D6-33A531024639](https://user-images.githubusercontent.com/44526468/102576315-35c78280-40c3-11eb-9333-a187e0a7cbe4.jpeg)

Eso lo hace generalmente cuando la conexión es un poco lenta, parece que la QString stdoutString cuando lee lo que envía picta-dl si demora mucho en obtener la info y como lo hace por un connect no lo recoge, incluso la misma salida del final que debe decir 100 % que para eso le hice un if

**Reply by Spheres-cu** [reply]
Así  se ve el gif en modo oscuro
![gif-modo-oscuro](https://user-images.githubusercontent.com/72468068/102644128-4b25c680-412e-11eb-9483-ca9fa0542746.png)

**Reply by Spheres-cu** [reply]
En el modo claro

![gif-modo-claro](https://user-images.githubusercontent.com/72468068/102644537-fb93ca80-412e-11eb-9462-529cbe697623.png)

**Reply by oleksis** [reply]
> Eso lo hace generalmente cuando la conexión es un poco lenta, parece que la QString stdoutString cuando lee lo que envía picta-dl si demora mucho en obtener la info y como lo hace por un connect no lo recoge, incluso la misma salida del final que debe decir 100 % que para eso le hice un if

Si eso parece; un caso es cuando la conectividad esta lenta y parece ser un comportamiento como describes.

**Reply by oleksis** [reply]
> Si ya el problema se le quitó, parece que era una función nueva que tuve que hacer para la animación con PNG y no con Gif ya que los gif en modo oscuro se ven mal, 

Si eso noto, parecen pincelados y la transparencia no trabaja del todo Ok 😕

> estoy tratando de hacer una función de animación para corregir eso, de lo contrario un gif para modo oscuro y otro para modo claro

Como más fácil te sea y en futuras versiones mejorar en cuanto a calidad

**Reply by Spheres-cu** [reply]
> Te envío una captura. De paso a veces el porciento de descarga cuando finaliza no reporta el 100%

Estoy haciendo unas pruebas pasándole el parámetro --newline a las descargas para emita el % de descarga en una linea nueva cada vez y a ver si asi no se escapada la salida del 100 %

**Comment by Spheres-cu** [comment]
Por ultimo los efectos de winApp.qss Vs el estilo semitransparente del command_linkButton
Con el winApp.qss

![efectos](https://user-images.githubusercontent.com/72468068/102645555-a5c02200-4130-11eb-8cf2-7255537b162c.png)

**Reply by Spheres-cu** [reply]
Sin el winApps.qss

![efectos2](https://user-images.githubusercontent.com/72468068/102646199-a5745680-4131-11eb-82ac-4edb400fae64.png)

**Reply by oleksis** [reply]
> Sin el winApps.qss
> 
> ![efectos2](https://user-images.githubusercontent.com/72468068/102646199-a5745680-4131-11eb-82ac-4edb400fae64.png)

Si trabaja mejor sin aplicar el estilo modifícalo en el main.cpp no incluirlo y quitarlo como recurso. Haz el cambio que mejor trabaje y las modificaciones a las Animaciones.

**Reply by rsanjuan87** [reply]
como decia en el telegram
el tema oscuro creo deba ser un +
lo veo bien

**Reply by oleksis** [reply]
Acabo de probar con winApp.qss como slylesheet en los recursos y asi se ve:

![GUI_con_winApp qss](https://user-images.githubusercontent.com/44526468/102658611-e1191b80-4145-11eb-9b90-936eeee58496.png)

Por qui todo bien!

**Reply by oleksis** [reply]
El otro detalle que faltaria por trabajar es la animacion de los iconos que como ves en la captura la transparencia/calidad no es la óptima.

**Comment by Spheres-cu** [comment]
Necesito una ayuda: encontré una forma de hacer animaciones basadas en PNG usando el formato APNG el cual al igual que el gif permite tener en un archivo una secuencia de imágenes para hacer una animación pero en este caso soportando las transparencias cosa que el gif no hace y por tanto en fondos oscuros se ve distorsionado si la base del gif es de color claro, el problema es que Qt no tiene soporte nativo para APNG pero existe un proyecto en Github que lo tienen implementado para Linux, Mac y Windows, el problema es que siguiendo los pasados que da el desarrollador no logro agregar su repositorio a los sources de Qt, yo tengo el Qt 5.9 y pienso migrar al 5.12 pero necesito probar si el problema es ese o es que algo no estoy haciendo bien, aqui les dejo los enlaces al proyecto de APNG:

https://github.com/Skycoder42/QtApng

https://github.com/Skycoder42/QtModules/blob/master/README.md#add-my-repositories-to-qt-maintenancetool

**Reply by oleksis** [reply]
> Necesito una ayuda: encontré una forma de hacer animaciones basadas en PNG usando el formato APNG el cual al igual que el gif permite tener en un archivo una secuencia de imágenes para hacer una animación pero en este caso soportando las transparencias cosa que el gif no hace y por tanto en fondos oscuros se ve distorsionado si la base del gif es de color claro, el problema es que Qt no tiene soporte nativo para APNG pero existe un proyecto en Github que lo tienen implementado para Linux, Mac y Windows, el problema es que siguiendo los pasados que da el desarrollador no logro agregar su repositorio a los sources de Qt, yo tengo el Qt 5.9 y pienso migrar al 5.12 pero necesito probar si el problema es ese o es que algo no estoy haciendo bien, aqui les dejo los enlaces al proyecto de APNG:
> 
> https://github.com/Skycoder42/QtApng
> 
> https://github.com/Skycoder42/QtModules/blob/master/README.md#add-my-repositories-to-qt-maintenancetool

Ok voy a probar/investigar sobre el tema.

@rsanjuan87

**Reply by oleksis** [reply]
@Spheres-cu Tienea algunos .gif que vamos a usar llevados a .apng? Ó convirtiendo algunos que tenemos a .apng hacemos las pruebas?

**Reply by Spheres-cu** [reply]
Tengo hecho algunas animaciones ya en apng, hay unas herramientas portables muy buenas para hacerlas, el tema es que Qt no las soporta a menos que se instale el plugin.

Una muestra de una animación apng
 
![download_button_grey_apng](https://user-images.githubusercontent.com/72468068/102703657-b27f6b80-423f-11eb-90e8-57162b9c5bde.png)

**Reply by Spheres-cu** [reply]
En modo oscuro

![download_button_black_apng](https://user-images.githubusercontent.com/72468068/102703676-eb1f4500-423f-11eb-9b12-0f57cc133d4a.png)

**Reply by Spheres-cu** [reply]
de todos modos hacerlas en modo oscuro no es necesario si se pudiera poner en apng ya que con la transparencia es suficiente para los dos modos

**Reply by oleksis** [reply]
> de todos modos hacerlas en modo oscuro no es necesario si se pudiera poner en apng ya que con la transparencia es suficiente para los dos modos

Por lo que estuve leyendo, instalar el qmodule/plugin dependeríamos de librería de terceros (además de la plataforma/arquitectura) al menos que se especifique en el proyecto (.pro) y la agreguemos localmente para cuando se compile desde GitHub en el Release sea más fácil.

Por ahora, mi recomendación es hacer dos versiones de los .gif animados para los temas claro y oscuro (light/dark) y por default cargar el tema claro, de especificar lo contrario establecer el tema oscuro. El sufijo del .gif puede ser _dark/light.gif y establecerlo en el widget mediante una variable que controle el tema habilitado. 

En futuras versiones podemos ver mejor esta integración usando APNG.

@rsanjuan87

**Comment by Spheres-cu** [comment]
En cuanto tenga las animaciones te las subo en un rar junto con los cambios de los formularios de la ayuda y el principal ya que tuve q hacer arreglos en el Layout para poner los iconos nuevos

**Reply by oleksis** [reply]
Ok bro! Estaré al tanto de la actualización para integrarlo/probar y hacer un nuevo release con los cambios. Release Christmas 🎄 🥳🎉

**Reply by Spheres-cu** [reply]
Quieres que le ponga Snow flakes en la GUI por navidad ? 😁🎉

**Reply by oleksis** [reply]
> Quieres que le ponga Snow flakes en la GUI por navidad ? 😁🎉

No es mala idea 💡 Estaba pensando que que se podría usar los emoticons 🎄🎉🥳 de navidad o fin de año y al lado del nombre (izquierda superior) después del logo poner un label y validar la fecha (la semana) para navidad 🎄 y para cuando sea 1 enero (nuevo año) la primera semana 🎉🎊 poner Felicidades . Algo como un huevo de pascuas . Sería genial 😁

**Comment by Spheres-cu** [comment]
Changelog:

- Agregado tema oscuro a la GUI.

- Cambio de los iconos para el tema oscuro o claro según lo tenga configurado en su sistema.

- Agregado el parámetro --newline para mejorar la obtención del progreso de descarga.

- ¡Feliz Navidad y Año Nuevo! 🎉❄️❄️❄️

[Cambios_interface_picta-dl-GUI.zip](https://github.com/oleksis/picta-dl-gui/files/5727399/Cambios_interface_picta-dl-GUI.zip)

**Reply by oleksis** [reply]
Que bien 😊 Probando e Integrando. Feliz Navidad 🎄 y Año nuevo 🎉🥳🎊

**Comment by oleksis** [comment]
@Spheres-cu 
> Otra cosa, en el tema oscuro actual cuando los botones estan desactivados pone fondo gris y en el gif se muestra una sombra clara, a ver si a esos botones se le puede poner en el estilo fondo oscuro para cuando estén desctivados y no se vea ese efecto

Si eso podríamos mejorarlo y ponerlos de acuerdo en cual color para el tema oscuro usar y establecerlo en los .gif

Ver que hexadecimal utilizamos por ejemplo:

```css
background-color: #001133;/*back*/
```
Por un dark común

El color común black para el tema oscuro es #0a0a0a

Sobre este tema es el comportamiento de los botones desactivados (gray out) que muestra un gris claro por encima para dar el efecto de desactivado. Debemos investigar una vía de hacer la animación (con .gif ó .apng) y si es posible a la vez no deshabilitar los botones cuando se estén animando.

Por ahora prefería seguir usando los .gif y trabajar sobre esta idea inicial. Establecer la animación desde que chequea si el programa está disponible al igual que cuando procesa la URL pero sin deshabilitar el CommandLinkButton

Habría que manejar el evento click para que cuando esté la animación activa ignorar el evento y las acciones de checque/proceso/descarga

Pienso que así evitaríamos el efecto del color gris al deshabilitar el botón

¿Que opinan al respecto?

@rsanjuan87

**Reply by rsanjuan87** [reply]
el qss lo ajuste a lo q pude del proyecto PsiphonQt, 
todos los cambios q crean necesarios, creo ustedes son los mas indicados para determinar q va mejor y mas incorporado al OS (MS Windows) pues yo no lo uso con frecuencia
